### PR TITLE
Sorting Cleanup

### PR DESCRIPTION
**Pull request by sslattery**
This PR does the following:

- [x] `LinkedCellList` becomes its own object. It has its own associated implementation of `permute` for `AoSoA` objects.

- [x] Improves the testing of `LinkedCellList`

- [x] Templates sorting data on the Cabana memory spaces rather than kokkos memory spaces.

- [x] Cleans the sort interface functions names for consistency.

### CODE REVIEW

**Comment by junghans** [comment]
Hmm, I still don't understand why the gitlab CI isn't triggered, most have something to do with the fact that the PR is coming for a fork, I guess.

File: core/src/Cabana_Sort.hpp [hunk 1]
```diff
@@ -114,173 +98,26 @@ class BinningData
...
 
 namespace Impl
 {
 //---------------------------------------------------------------------------//
 // Create a permutation vector over a range subset using a comparator over the
 // given Kokkos View of keys.
 template<class KeyViewType, class Comparator>
-BinningData<typename KeyViewType::memory_space>
+BinningData<
+    typename KokkosSpaceToCabana<typename KeyViewType::memory_space>::type>
```
Comment (rfbird) [inline core/src/Cabana_Sort.hpp]: That is a really nice enhancement, good stuff

File: core/src/impl/Cabana_CartesianGrid.hpp [hunk 2]
```diff
@@ -20,52 +20,62 @@ namespace Cabana
...
-    Scalar _max_x;
-    Scalar _max_y;
-    Scalar _max_z;
-    Scalar _dx;
-    Scalar _dy;
-    Scalar _dz;
-    Scalar _rdx;
-    Scalar _rdy;
-    Scalar _rdz;
+    double _min_x;
```
Comment (rfbird) [inline core/src/impl/Cabana_CartesianGrid.hpp]: As per my comment on Slack, I don't really like templating on data stype..but not do I like being married down to double (I think it makes this Cartesian Grid less generic, even though it makes sense for the Vertlet being the user)

File: core/src/impl/Cabana_CartesianGrid.hpp [hunk 3]
```diff
@@ -75,11 +85,25 @@ class CartesianGrid
         num_z = _nz;
     }
 
+    // Get the number of cells in a given direction.
+    KOKKOS_INLINE_FUNCTION
```
Comment (rfbird) [inline core/src/impl/Cabana_CartesianGrid.hpp]: Did we want to wrap this `Kokkos_xxxx` with `Cabana_xxxx`?

File: core/unit_test/tstSort.hpp [hunk 4]
```diff
@@ -385,7 +385,7 @@ void testBinByMemberDataOnly()
     // Bin the aosoa by the 1D member. Use one bin per data point to
     // effectively make this a sort. Don't actually move the particle data
     // though - just create the binning data.
-    auto bin_data = Cabana::binByMember( aosoa.slice<1>(), num_data-1 );
+    auto bin_data = Cabana::binByKey( aosoa.slice<1>(), num_data-1 );
```
Comment (rfbird) [inline core/unit_test/tstSort.hpp]: Is this planned to be the last version of the `sort` interface before we lock it down?

Comment (rfbird) [reply core/unit_test/tstSort.hpp]: (not that I have a problem with it, I just hope to play around with it a bit for my PIC use case if possible before marrying down to it. As per our last discussion on this I think we had my requirements captured)

File: core/src/impl/Cabana_CartesianGrid.hpp (hunk 3 above)
Comment (sslattery) [reply core/src/impl/Cabana_CartesianGrid.hpp]: I'm torn if we should propagate the Cabana macros to implementation details

File: core/unit_test/tstSort.hpp (hunk 4 above)
Comment (sslattery) [reply core/unit_test/tstSort.hpp]: This PR is hopefully the last update to the code API before tagging a release candidate. This captures your requirements I think but I wanted to make it more consistent in terms of naming convention.

**Pull request merged**